server/ServerApp.cpp

```diff
@@ -2288,36 +2288,42 @@ namespace {
         }
     }
 
-    void GetFleetsVisibleToEmpireAtSystem(std::set<int>& visible_fleets, int empire_id, int system_id,
-                                          const ScriptingContext& context)
+    bool HasBlockadeOrGetFleetsVisibleToEmpireAtSystem(std::set<int>& visible_fleets, int empire_id,
```

Comment:
this is not a good function design... it's a weird mix of return value and in/out parameters, and it's very confusingly doing... something? name not at all clear.

Reply:
also, the name "HasBlockade" is unclear... the empire_id has a ship that is blockaded, or that empire has a ship that is blockading something else there?

Reply:
are we at C++17 ? we could probably return `std::variant<std::set<int>, boolean>` , the intention here is that only one of the results is meaningful. Actually it should be either a set or a signal that there is a blockade (in which case visibility does not matter in this case).

we certainly could also do an extra pass just to check for blockade, but that seems unnecessarily wasteful.

An old school recommendation seems: always return out params as return value and put those into a kind of structure when there are multiple out params
according to https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out-multi

Reply:
I am not sure if this is possible, but something like the following would be my preferred signature (in pseudo c++):
```std::variant<std::set<int>, ProcessingShortcut::AbortedDueToBlockadeAtSystem> GetFleetsVisibleToEmpireAtSystem(int empire_id, int system_id, const ScriptingContext& context)```
The function name would explain the standard return value (a set of fleets) and the enum value would describe the non-standard/abort result.

Reply:
> are we at C++17 ?

In FreeOrion code (not GiGi code), standard is C++20

> we could probably return `std::variant<std::set<int>, boolean>` , the intention here is that only one of the results is meaningful. Actually it should be either a set or a signal that there is a blockade (in which case visibility does not matter in this case).

Could work, but I think just pair<set<int>, bool> would be fine. The set can be ignored if not needed.

> we certainly could also do an extra pass just to check for blockade, but that seems unnecessarily wasteful.

I'd rather do that: separate functions for what is visible at a system to an empire, and a check if there is a blockade at a system.